pkg/segment/query/iqr/intermediateQueryResult.go

```diff
@@ -0,0 +1,372 @@
+// Copyright (c) 2021-2024 SigScalr, Inc.
+//
+// This file is part of SigLens Observability Solution
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package iqr
+
+import (
+	"fmt"
+
+	"github.com/siglens/siglens/pkg/ast/pipesearch"
+	"github.com/siglens/siglens/pkg/segment/reader/record"
+	"github.com/siglens/siglens/pkg/segment/utils"
+	toputils "github.com/siglens/siglens/pkg/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+type iqrMode int
+
+const (
+	invalidMode iqrMode = iota
+	notSet
+	withRRCs
+	withoutRRCs
+)
+
+type IQR struct {
+	mode iqrMode
+
+	// Used if and only if the mode is withRRCs.
+	rrcs             []*utils.RecordResultContainer
+	encodingToSegKey map[uint16]string
+
+	// Used in both modes.
+	qid            uint64
+	knownValues    map[string][]utils.CValueEnclosure
+	deletedColumns map[string]struct{}
+	renamedColumns map[string]string
+
+	// Used only if the mode is withoutRRCs. Sometimes not used in that mode.
+	groupbyColumns []string
+	measureColumns []string
+}
+
+func NewIQR(qid uint64) *IQR {
+	return &IQR{
+		mode:             notSet,
+		qid:              qid,
+		rrcs:             make([]*utils.RecordResultContainer, 0),
+		encodingToSegKey: make(map[uint16]string),
+		knownValues:      make(map[string][]utils.CValueEnclosure),
+		deletedColumns:   make(map[string]struct{}),
+		renamedColumns:   make(map[string]string),
+		groupbyColumns:   make([]string, 0),
+		measureColumns:   make([]string, 0),
+	}
+}
+
+func (iqr *IQR) validate() error {
+	if iqr == nil {
+		return fmt.Errorf("IQR is nil")
+	}
+
+	if iqr.mode == invalidMode {
+		return fmt.Errorf("IQR.mode is invalid")
+	}
+
+	for cname := range iqr.knownValues {
+		if len(iqr.knownValues[cname]) != len(iqr.rrcs) && len(iqr.rrcs) != 0 {
```

Comment:
nit: not sure how go compiler does this under the hood but was thinking:
we should do `for cname, cvals := range ...` 
and also `lenOfCvals := len(cvals)` so that we can reuse these two vars in below checks and avoid the second map lookup len call.

Reply:
actually nvm, the second call for map lookup and len calculations happens only in error scenario

Reply:
Yes I updated this to `for cname, values := range... ` because I think that's better. I didn't make the `lendOfValues` variable though because it only happens in the error case and `len` should be an O(1) operation anyways, so I doubt there'd be much/any performance change